### ISSUE TITLE
Bug with local course file missing "language"

### ISSUE THREAD

**Issue by nikke234** [Aalto-LeTech/intellij-plugin#315]
A test user had the following errors when using _Turn Project Into A+ Course Project_:
![stacktrace1](https://user-images.githubusercontent.com/30376763/90139041-81b69380-dd80-11ea-98ae-e6405075a029.png)
![stacktrace2](https://user-images.githubusercontent.com/30376763/90139037-811dfd00-dd80-11ea-841f-22ed83fc8b24.png)

Weirdly enough, as seen in the second stack trace, the course project action called `CourseFileManager#load`, which should only happen if the course file exists already. The course file JSON had just the URL and no modules or language. The project was a Scala SBT project instead of an empty project. The user was not able to reproduce the issue.

**Comment by superseacat** [comment]
Ubuntu

```
org.json.JSONException: JSONObject["exercise_info"] is not a JSONObject.
	at org.json.JSONObject.getJSONObject(JSONObject.java:770)
	at fi.aalto.cs.apluscourses.model.SubmissionInfo.fromJsonObject(SubmissionInfo.java:35)
	at fi.aalto.cs.apluscourses.dal.APlusExerciseDataSource$DefaultDataAccess.parseSubmissionInfo(APlusExerciseDataSource.java:219)
	at fi.aalto.cs.apluscourses.dal.APlusExerciseDataSource.getSubmissionInfo(APlusExerciseDataSource.java:78)
	at fi.aalto.cs.apluscourses.intellij.actions.SubmitExerciseAction.trySubmit(SubmitExerciseAction.java:152)
	at fi.aalto.cs.apluscourses.intellij.actions.SubmitExerciseAction.actionPerformed(SubmitExerciseAction.java:120)
	at com.intellij.openapi.actionSystem.ex.ActionUtil.performActionDumbAware(ActionUtil.java:282)
	at com.intellij.openapi.actionSystem.ex.ActionUtil.performActionDumbAwareWithCallbacks(ActionUtil.java:276)
	at com.intellij.openapi.actionSystem.impl.ActionButton.actionPerformed(ActionButton.java:182)
	at com.intellij.openapi.actionSystem.impl.ActionButton.performAction(ActionButton.java:155)
	at com.intellij.openapi.actionSystem.impl.ActionButton.processMouseEvent(ActionButton.java:439)
	at java.desktop/java.awt.Component.processEvent(Component.java:6415)
	at java.desktop/java.awt.Container.processEvent(Container.java:2263)
	at java.desktop/java.awt.Component.dispatchEventImpl(Component.java:5025)
	at java.desktop/java.awt.Container.dispatchEventImpl(Container.java:2321)
	at java.desktop/java.awt.Component.dispatchEvent(Component.java:4857)
	at java.desktop/java.awt.LightweightDispatcher.retargetMouseEvent(Container.java:4918)
	at java.desktop/java.awt.LightweightDispatcher.processMouseEvent(Container.java:4547)
	at java.desktop/java.awt.LightweightDispatcher.dispatchEvent(Container.java:4488)
	at java.desktop/java.awt.Container.dispatchEventImpl(Container.java:2307)
	at java.desktop/java.awt.Window.dispatchEventImpl(Window.java:2773)
	at java.desktop/java.awt.Component.dispatchEvent(Component.java:4857)
	at java.desktop/java.awt.EventQueue.dispatchEventImpl(EventQueue.java:778)
	at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:727)
	at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:721)
	at java.base/java.security.AccessController.doPrivileged(Native Method)
	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:85)
	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:95)
	at java.desktop/java.awt.EventQueue$5.run(EventQueue.java:751)
	at java.desktop/java.awt.EventQueue$5.run(EventQueue.java:749)
	at java.base/java.security.AccessController.doPrivileged(Native Method)
	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:85)
	at java.desktop/java.awt.EventQueue.dispatchEvent(EventQueue.java:748)
	at com.intellij.ide.IdeEventQueue.defaultDispatchEvent(IdeEventQueue.java:967)
	at com.intellij.ide.IdeEventQueue.dispatchMouseEvent(IdeEventQueue.java:904)
	at com.intellij.ide.IdeEventQueue._dispatchEvent(IdeEventQueue.java:836)
	at com.intellij.ide.IdeEventQueue.lambda$dispatchEvent$8(IdeEventQueue.java:450)
	at com.intellij.openapi.progress.impl.CoreProgressManager.computePrioritized(CoreProgressManager.java:744)
	at com.intellij.ide.IdeEventQueue.lambda$dispatchEvent$9(IdeEventQueue.java:449)
	at com.intellij.openapi.application.impl.ApplicationImpl.runIntendedWriteActionOnCurrentThread(ApplicationImpl.java:802)
	at com.intellij.ide.IdeEventQueue.dispatchEvent(IdeEventQueue.java:503)
	at java.desktop/java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:203)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:124)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:113)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:109)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:101)
	at java.desktop/java.awt.EventDispatchThread.run(EventDispatchThread.java:90)`

**Comment by superseacat** [comment]
feels like something is fishy with A+'s API (or auth)

**Comment by superseacat** [comment]
```
org.json.JSONException: JSONObject["language"] not found.
	at org.json.JSONObject.get(JSONObject.java:573)
	at org.json.JSONObject.getString(JSONObject.java:836)
	at fi.aalto.cs.apluscourses.intellij.utils.CourseFileManager.loadFromJsonObject(CourseFileManager.java:184)
	at fi.aalto.cs.apluscourses.intellij.utils.CourseFileManager.load(CourseFileManager.java:79)
	at fi.aalto.cs.apluscourses.intellij.utils.CourseFileManager.createAndLoad(CourseFileManager.java:51)
	at fi.aalto.cs.apluscourses.intellij.actions.CourseProjectAction.tryCreateCourseFile(CourseProjectAction.java:265)
	at fi.aalto.cs.apluscourses.intellij.actions.CourseProjectAction.actionPerformed(CourseProjectAction.java:155)
```

**Comment by superseacat** [comment]
idea #1:

so let's assume, that on some platforms (1 user reported on macOS, and one intern-tester Ubuntu), this
https://github.com/Aalto-LeTech/intellij-plugin/blob/7862398d2cb26c8d7f2df00e2a50d686fcb17da4/src/main/java/fi/aalto/cs/apluscourses/intellij/utils/CourseFileManager.java#L77
does not always work, because of

- https://stackoverflow.com/questions/31350336/within-intellij-ide-the-java-file-exists-always-returns-false
- https://stackoverflow.com/questions/33271744/intellij-filenotfoundexception-file-exists/48958694

for the reasons described. could this theoretically explain the 

> Weirdly enough, as seen in the second stack trace, the course project action called CourseFileManager#load, which should only happen if the course file exists already.

@nikke234 ?

**Comment by superseacat** [comment]
idea # 2:

so the NPE happens on "language", but there's one call already before that:
https://github.com/Aalto-LeTech/intellij-plugin/blob/7862398d2cb26c8d7f2df00e2a50d686fcb17da4/src/main/java/fi/aalto/cs/apluscourses/intellij/utils/CourseFileManager.java#L182-L184

could this mean, that all the previous logics (file does exist, and contains a URL too) and the problem is in the language parameter not being written quickly enough into the file?

in other words, "turn into A+ dialog" prompts the user a language, and gets it, but tries to read the "language" parameter of `a-plus-project.json` before it arrives (gets written) there?

**Comment by superseacat** [comment]
**this is getting separated definitely:** https://github.com/Aalto-LeTech/intellij-plugin/blob/3c58bc0d75c5df220e935bc9ea3ca3e8698710d9/src/main/java/fi/aalto/cs/apluscourses/intellij/activities/InitializationActivity.java#L80

**Comment by superseacat** [comment]
```
org.json.JSONException: JSONObject["test"] not found.
	at org.json.JSONObject.get(JSONObject.java:573)
	at org.json.JSONObject.getString(JSONObject.java:836)
	at fi.aalto.cs.apluscourses.intellij.utils.CourseFileManager.loadFromJsonObject(CourseFileManager.java:184)
	at fi.aalto.cs.apluscourses.intellij.utils.CourseFileManager.load(CourseFileManager.java:79)
	at fi.aalto.cs.apluscourses.intellij.utils.CourseFileManager.createAndLoad(CourseFileManager.java:51)
	at fi.aalto.cs.apluscourses.intellij.actions.CourseProjectAction.tryCreateCourseFile(CourseProjectAction.java:265)
	at fi.aalto.cs.apluscourses.intellij.actions.CourseProjectAction.actionPerformed(CourseProjectAction.java:155)`

**Comment by stellatsv** [comment]
Most things relating to this have been changed since.
